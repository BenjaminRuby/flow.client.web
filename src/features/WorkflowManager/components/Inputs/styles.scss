@import "~@boomerang/boomerang-styles/lib/colors";

.c-workflow-inputs {
  display: flex;
  flex-flow: row wrap;
  padding: 2rem;
}

.b-workflow-input {
  position: relative;
  background-color: $white;
  min-height: 25rem;
  width: 20rem;
  padding: 1rem;
  margin: 0 2rem 2rem 0;
  border-left: 0.3125rem solid;
  border-radius: 0.25rem;

  &.--text {
    border-left-color: $green-30;
  }
  &.--textarea {
    border-left-color: $ultramarine-30;
  }
  &.--boolean {
    border-left-color: $red-30;
  }
  &.--password {
    border-left-color: $gray-30;
  }
  &.--number {
    border-left-color: $gold-30;
  }
  &.--select {
    border-left-color: $violet-30;
  }
}

.b-workflow-input__name {
  font-size: 1.25rem;
  font-weight: 600;
  letter-spacing: 0.0625rem;
  text-overflow: ellipsis;
  overflow: hidden;
  margin: 0;
  margin-bottom: 1rem;
  line-height: 1.5rem;
}

.b-workflow-input-field {
  display: flex;
  flex-direction: column;
  font-size: 1rem;
  text-overflow: ellipsis;
  overflow: hidden;
  margin-top: 1rem;
}

.b-workflow-input-field__key {
  font-weight: 600;
  line-height: 1.125rem;
  margin-bottom: 0.75rem;
  text-overflow: ellipsis;
  overflow: hidden;
  white-space: nowrap;
}

.b-workflow-input-field__value {
  font-weight: 400;
  text-align: left;
  line-height: 1.25rem;
  text-overflow: ellipsis;
  overflow: hidden;
  white-space: nowrap;
}

.b-workflow-input__delete {
  position: absolute;
  top: 1rem;
  right: 1rem;
  cursor: pointer;
}

.b-workflow-input-edit {
  position: absolute;
  display: flex;
  align-items: center;
  font-size: 1rem;
  font-weight: 600;
  bottom: 1rem;
  right: 1rem;
  cursor: pointer;
}

.b-workflow-input-edit__pencil {
  height: 0.8rem;
  margin-left: 0.5rem;
}

.b-workflow-input-create {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  min-height: 25rem;
  height: calc(100% - 1.75rem);
  width: 20rem;
  border: 0.0625rem dashed $gray-30;
  color: $gray-30;
  cursor: pointer;
  border-radius: 0.25rem;
}

.b-workflow-input-create__plus {
  margin-bottom: 2rem;
}
