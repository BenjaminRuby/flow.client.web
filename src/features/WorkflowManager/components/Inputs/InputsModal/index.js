import React, { Component } from "react";
import PropTypes from "prop-types";
import { ModalFlow } from "@boomerang/carbon-addons-boomerang-react";
import { Add32, Edit32 } from "@carbon/icons-react";
import InputsModalContent from "./InputsModalContent";
import "./styles.scss";

class InputsModal extends Component {
  static propTypes = {
    isEdit: PropTypes.bool.isRequired,
    loading: PropTypes.bool.isRequired,
    inputsKeys: PropTypes.array,
    input: PropTypes.object,
    updateInputs: PropTypes.func.isRequired
  };

  editTrigger = ({ openModal }) => {
    let output = null;
    this.props.isEdit
      ? (output = (
          <button className="b-workflow-input-edit" onClick={openModal}>
            Edit
            <Edit32 className="b-workflow-input-edit__pencil" />
          </button>
        ))
      : (output = (
          <div className="b-workflow-input-create">
            <button className="b-workflow-input-create__plus" onClick={openModal}>
              <Add32 className="b-workflow-input-create__plus-icon" />
            </button>
            Create New Property
          </div>
        ));
    return output;
  };

  render() {
    const { isEdit, input, loading } = this.props;

    return (
      <ModalFlow
        composedModalProps={{ containerClassName: "c-inputs-modal" }}
        confirmModalProps={{
          title: "Close this?",
          children: <div>Your input will not be saved</div>
        }}
        modalHeaderProps={{
          title: isEdit && input ? input.label : "Create Property",
          subTitle: isEdit ? "Let's update it" : "Create new input property"
        }}
        modalTrigger={this.editTrigger}
      >
        <InputsModalContent loading={loading} updateInputs={this.props.updateInputs} {...this.props} />
      </ModalFlow>
    );
  }
}

export default InputsModal;
