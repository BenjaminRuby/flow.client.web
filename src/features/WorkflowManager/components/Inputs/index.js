import React, { Component } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { bindActionCreators } from "redux";
import { actions as workflowActions } from "State/workflow";
import classnames from "classnames";
import AlertModalWrapper from "@boomerang/boomerang-components/lib/AlertModal";
import ConfirmModal from "@boomerang/boomerang-components/lib/ConfirmModal";
import Tooltip from "@boomerang/boomerang-components/lib/Tooltip";
import InputsModal from "./InputsModal";
import { Add32, Close32, Edit32 } from "@carbon/icons-react";
import INPUT_TYPES from "Constants/workflowInputTypes";
import "./styles.scss";

class Inputs extends Component {
  static propTypes = {
    inputs: PropTypes.array.isRequired,
    loading: PropTypes.bool.isRequired,
    updateInputs: PropTypes.func.isRequired,
    workflowActions: PropTypes.object.isRequired
  };

  formatDefaultValue = value => {
    switch (value) {
      case INPUT_TYPES.BOOLEAN:
        return value.toString();
      default:
        return value;
    }
  };

  deleteInput = key => {
    new Promise(resolve => {
      resolve(this.props.workflowActions.deleteWorkflowInput({ key }));
    }).then(() =>
      this.props.updateInputs({ title: "Delete Input", message: "Successfully deleted input", type: "delete" })
    );
  };

  render() {
    const { inputs } = this.props;
    const inputsKeys = inputs.map(input => input.key);
    return (
      <div className="c-workflow-inputs">
        {inputs.length > 0 &&
          inputs.map((input, index) => (
            <div
              key={`${input.id}-${index}`}
              className={classnames("b-workflow-input", `--${input.type}`, `--${input.readOnly}`)}
            >
              <h1 className="b-workflow-input__name">{input.label}</h1>
              <dl className="b-workflow-input-field">
                <dt className="b-workflow-input-field__key">Key</dt>
                <dd className="b-workflow-input-field__value">{input.key}</dd>
              </dl>
              <dl className="b-workflow-input-field">
                <dt className="b-workflow-input-field__key">Description</dt>
                <dd className="b-workflow-input-field__value">{input.description}</dd>
              </dl>
              <dl className="b-workflow-input-field">
                <dt className="b-workflow-input-field__key">Type</dt>
                <dd className="b-workflow-input-field__value">{input.type}</dd>
              </dl>
              <dl className="b-workflow-input-field">
                <dt className="b-workflow-input-field__key">Required</dt>
                <dd className="b-workflow-input-field__value">{input.required.toString()}</dd>
              </dl>
              <dl className="b-workflow-input-field">
                <dt className="b-workflow-input-field__key">Default value</dt>
                <dd className="b-workflow-input-field__value">{this.formatDefaultValue(input.defaultValue)}</dd>
              </dl>
              {input.validValues && (
                <dl className="b-workflow-input-field">
                  <dt className="b-workflow-input-field__key">Valid values</dt>
                  <dd className="b-workflow-input-field__value">
                    {this.formatDefaultValue(input.validValues.join(", "))}
                  </dd>
                </dl>
              )}
              {!input.readOnly ? (
                <AlertModalWrapper
                  ModalTrigger={() => (
                    <button className="b-workflow-input__delete">
                      <Close32 data-tip data-for={`${input.id}`} className="b-workflow-input__delete-icon" />
                      <Tooltip id={`${input.id}`} place="top">
                        Delete Input
                      </Tooltip>
                    </button>
                  )}
                  modalContent={(closeModal, rest) => (
                    <ConfirmModal
                      closeModal={closeModal}
                      affirmativeAction={() => {
                        closeModal();
                        this.deleteInput(input.key);
                      }}
                      title="DELETE THIS PROPERTY?"
                      subTitleTop="It will be gone. Forever."
                      cancelText="NO"
                      affirmativeText="YES"
                      theme="bmrg-flow"
                      {...rest}
                    />
                  )}
                />
              ) : null}
              {!input.readOnly ? (
                <InputsModal
                  isEdit
                  inputsKeys={inputsKeys.filter(inputName => inputName !== input.key)}
                  Button={() => (
                    <button className="b-workflow-input-edit">
                      Edit
                      <Edit32 className="b-workflow-input-edit__pencil" />
                    </button>
                  )}
                  input={input}
                  updateInputs={this.props.updateInputs}
                  loading={this.props.loading}
                />
              ) : null}
            </div>
          ))}
        <InputsModal
          isEdit={false}
          inputsKeys={inputsKeys}
          Button={() => (
            <div className="b-workflow-input-create">
              <button className="b-workflow-input-create__plus">
                <Add32 className="b-workflow-input-create__plus-icon" />
              </button>
              Create New Property
            </div>
          )}
          updateInputs={this.props.updateInputs}
          loading={this.props.loading}
        />
      </div>
    );
  }
}

const mapStateToProps = state => ({
  inputs: state.workflow.data.properties
});

const mapDispatchToProps = dispatch => ({
  workflowActions: bindActionCreators(workflowActions, dispatch)
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Inputs);
