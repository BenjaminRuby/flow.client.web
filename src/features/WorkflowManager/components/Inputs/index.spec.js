import React from "react";
import Inputs from ".";
import { fireEvent } from "@testing-library/react";

const mockfn = jest.fn();

const initialState = {
  workflow: {
    data: {
      properties: [
        {
          defaultValue: "pandas",
          description: "Tim property",
          key: "tim.property",
          label: "Tim Property",
          required: true,
          type: "select",
          validValues: ["pandas", "dogs"]
        }
      ]
    }
  }
};

const props = {
  loading: false,
  updateInputs: mockfn,
  workflowActions: { deleteWorkflowInput: mockfn }
};

beforeEach(() => {
  document.body.setAttribute("id", "app");
});

describe("Inputs --- Snapshot Test", () => {
  it("Capturing Snapshot of Inputs", () => {
    const { baseElement } = rtlReduxRender(<Inputs {...props} />, { initialState });
    expect(baseElement).toMatchSnapshot();
  });
});

describe("Inputs --- RTL", () => {
  it("Render inputs correctly", () => {
    const { queryByText } = rtlReduxRender(<Inputs {...props} />, { initialState });
    expect(queryByText("tim.property")).toBeInTheDocument();
  });

  it("Opens create new property modal", () => {
    const { queryByText, getByTestId } = rtlReduxRender(<Inputs {...props} />, { initialState });

    //expect(queryByText(/Create new input property/i)).not.toBeInTheDocument();

    //const modalTrigger = getByText(/Create new property/i);
    const modalTrigger = getByTestId("create-new-workflow-input-button");
    fireEvent.click(modalTrigger);

    expect(queryByText(/Create new input property/i)).toBeInTheDocument();
  });

  it("Opens edit property modal", () => {
    const { getByText, queryByText } = rtlReduxRender(<Inputs {...props} />, { initialState });

    //expect(queryByText(/Let's update it/i)).not.toBeInTheDocument();

    const modalTrigger = getByText(/Edit/i);
    fireEvent.click(modalTrigger);

    expect(queryByText(/Let's update it/i)).toBeInTheDocument();
  });
});
