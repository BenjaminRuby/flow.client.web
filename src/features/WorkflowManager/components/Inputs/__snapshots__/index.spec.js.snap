// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Inputs --- Snapshot Test Capturing Snapshot of Inputs 1`] = `
<body
  id="app"
>
  <div>
    <div
      class="c-workflow-inputs"
    >
      <div
        class="b-workflow-input --select --undefined"
      >
        <h1
          class="b-workflow-input__name"
        >
          Tim Property
        </h1>
        <dl
          class="b-workflow-input-field"
        >
          <dt
            class="b-workflow-input-field__key"
          >
            Key
          </dt>
          <dd
            class="b-workflow-input-field__value"
          >
            tim.property
          </dd>
        </dl>
        <dl
          class="b-workflow-input-field"
        >
          <dt
            class="b-workflow-input-field__key"
          >
            Description
          </dt>
          <dd
            class="b-workflow-input-field__value"
          >
            Tim property
          </dd>
        </dl>
        <dl
          class="b-workflow-input-field"
        >
          <dt
            class="b-workflow-input-field__key"
          >
            Type
          </dt>
          <dd
            class="b-workflow-input-field__value"
          >
            select
          </dd>
        </dl>
        <dl
          class="b-workflow-input-field"
        >
          <dt
            class="b-workflow-input-field__key"
          >
            Required
          </dt>
          <dd
            class="b-workflow-input-field__value"
          >
            true
          </dd>
        </dl>
        <dl
          class="b-workflow-input-field"
        >
          <dt
            class="b-workflow-input-field__key"
          >
            Default value
          </dt>
          <dd
            class="b-workflow-input-field__value"
          >
            pandas
          </dd>
        </dl>
        <dl
          class="b-workflow-input-field"
        >
          <dt
            class="b-workflow-input-field__key"
          >
            Valid values
          </dt>
          <dd
            class="b-workflow-input-field__value"
          >
            pandas, dogs
          </dd>
        </dl>
        <button
          class="b-workflow-input__delete"
        >
          <svg
            aria-hidden="true"
            class="b-workflow-input__delete-icon"
            currentItem="false"
            data-for="undefined"
            data-tip="true"
            focusable="false"
            height="32"
            preserveAspectRatio="xMidYMid meet"
            style="will-change: transform;"
            viewBox="0 0 32 32"
            width="32"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M24 9.4L22.6 8 16 14.6 9.4 8 8 9.4l6.6 6.6L8 22.6 9.4 24l6.6-6.6 6.6 6.6 1.4-1.4-6.6-6.6L24 9.4z"
            />
          </svg>
          <div
            class="__react_component_tooltip place-top type-dark "
            data-id="tooltip"
            id="undefined"
          >
            Delete Input
          </div>
        </button>
        <div
          class="bx--modal bx--bmrg-modal-confirm-container"
          role="presentation"
          tabindex="-1"
        >
          <div
            class="bx--modal-container"
          >
            <div
              class="bx--modal-header"
            >
              <p
                class="bx--modal-header__label bx--type-delta"
              >
                It will be gone. Forever.
              </p>
              <p
                class="bx--modal-header__heading bx--type-beta"
              >
                DELETE THIS PROPERTY?
              </p>
              <button
                class="bx--modal-close"
                title="Close"
                type="button"
              >
                <svg
                  aria-label="Close"
                  class="bx--modal-close__icon"
                  focusable="false"
                  height="20"
                  preserveAspectRatio="xMidYMid meet"
                  role="img"
                  style="will-change: transform;"
                  viewBox="0 0 32 32"
                  width="20"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M24 9.4L22.6 8 16 14.6 9.4 8 8 9.4l6.6 6.6L8 22.6 9.4 24l6.6-6.6 6.6 6.6 1.4-1.4-6.6-6.6L24 9.4z"
                  />
                </svg>
              </button>
            </div>
            <div
              id="confirm-modal-container"
            >
              <div
                class="bx--modal-content"
              >
                <p
                  class="bx--modal-content__text"
                />
              </div>
            </div>
            <div
              class="bx--modal-footer"
              inputref="[object Object]"
            >
              <button
                class="bx--btn bx--btn--secondary"
                id="bx--bmrg-confirm-modal-secondary-button"
                tabindex="0"
                type="button"
              >
                NO
              </button>
              <button
                class="bx--btn bx--btn--primary"
                id="bx--bmrg-confirm-modal-primary-button"
                tabindex="0"
                type="button"
              >
                YES
              </button>
            </div>
          </div>
        </div>
        <button
          class="b-workflow-input-edit"
        >
          Edit
          <svg
            aria-hidden="true"
            class="b-workflow-input-edit__pencil"
            focusable="false"
            height="32"
            preserveAspectRatio="xMidYMid meet"
            style="will-change: transform;"
            viewBox="0 0 32 32"
            width="32"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M2 27h28v2H2zM25.41 9a2 2 0 0 0 0-2.83l-3.58-3.58a2 2 0 0 0-2.83 0l-15 15V24h6.41zm-5-5L24 7.59l-3 3L17.41 7zM6 22v-3.59l10-10L19.59 12l-10 10z"
            />
          </svg>
        </button>
        <div
          class="bx--modal bx--bmrg-modal-flow"
          role="presentation"
          tabindex="-1"
        >
          <div
            class="bx--modal-container bx--bmrg-modal-flow-container c-inputs-modal"
          >
            <div
              class="bx--modal-header"
              subtitle="Let's update it"
            >
              <p
                class="bx--modal-header__heading bx--type-beta"
              >
                TIM PROPERTY
              </p>
              <p
                class="bx--bmrg-modal-flow-subtitle"
              >
                Let's update it
              </p>
              <button
                class="bx--modal-close"
                title="Close the modal"
                type="button"
              >
                <svg
                  aria-label="Close the modal"
                  class="bx--modal-close__icon"
                  focusable="false"
                  height="20"
                  preserveAspectRatio="xMidYMid meet"
                  role="img"
                  style="will-change: transform;"
                  viewBox="0 0 32 32"
                  width="20"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M24 9.4L22.6 8 16 14.6 9.4 8 8 9.4l6.6 6.6L8 22.6 9.4 24l6.6-6.6 6.6 6.6 1.4-1.4-6.6-6.6L24 9.4z"
                  />
                </svg>
              </button>
            </div>
            <div
              class="bx--modal bx--bmrg-modal-confirm-container"
              role="presentation"
              tabindex="-1"
            >
              <div
                class="bx--modal-container"
              >
                <div
                  class="bx--modal-header"
                >
                  <p
                    class="bx--modal-header__heading bx--type-beta"
                  >
                    Close this?
                  </p>
                  <button
                    class="bx--modal-close"
                    title="Close"
                    type="button"
                  >
                    <svg
                      aria-label="Close"
                      class="bx--modal-close__icon"
                      focusable="false"
                      height="20"
                      preserveAspectRatio="xMidYMid meet"
                      role="img"
                      style="will-change: transform;"
                      viewBox="0 0 32 32"
                      width="20"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M24 9.4L22.6 8 16 14.6 9.4 8 8 9.4l6.6 6.6L8 22.6 9.4 24l6.6-6.6 6.6 6.6 1.4-1.4-6.6-6.6L24 9.4z"
                      />
                    </svg>
                  </button>
                </div>
                <div
                  id="confirm-modal-container"
                >
                  <div
                    class="bx--modal-content"
                  >
                    <p
                      class="bx--modal-content__text"
                    >
                      <div>
                        Your input will not be saved
                      </div>
                    </p>
                  </div>
                </div>
                <div
                  class="bx--modal-footer"
                  inputref="[object Object]"
                >
                  <button
                    class="bx--btn bx--btn--secondary"
                    id="bx--bmrg-confirm-modal-secondary-button"
                    tabindex="0"
                    type="button"
                  >
                    NO
                  </button>
                  <button
                    class="bx--btn bx--btn--primary"
                    id="bx--bmrg-confirm-modal-primary-button"
                    tabindex="0"
                    type="button"
                  >
                    YES
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="b-workflow-input-create"
      >
        <button
          class="b-workflow-input-create__plus"
        >
          <svg
            aria-hidden="true"
            class="b-workflow-input-create__plus-icon"
            focusable="false"
            height="32"
            preserveAspectRatio="xMidYMid meet"
            style="will-change: transform;"
            viewBox="0 0 32 32"
            width="32"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M17 15V7h-2v8H7v2h8v8h2v-8h8v-2h-8z"
            />
          </svg>
        </button>
        Create New Property
      </div>
      <div
        class="bx--modal bx--bmrg-modal-flow"
        role="presentation"
        tabindex="-1"
      >
        <div
          class="bx--modal-container bx--bmrg-modal-flow-container c-inputs-modal"
        >
          <div
            class="bx--modal-header"
            subtitle="Create new input property"
          >
            <p
              class="bx--modal-header__heading bx--type-beta"
            >
              CREATE PROPERTY
            </p>
            <p
              class="bx--bmrg-modal-flow-subtitle"
            >
              Create new input property
            </p>
            <button
              class="bx--modal-close"
              title="Close the modal"
              type="button"
            >
              <svg
                aria-label="Close the modal"
                class="bx--modal-close__icon"
                focusable="false"
                height="20"
                preserveAspectRatio="xMidYMid meet"
                role="img"
                style="will-change: transform;"
                viewBox="0 0 32 32"
                width="20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M24 9.4L22.6 8 16 14.6 9.4 8 8 9.4l6.6 6.6L8 22.6 9.4 24l6.6-6.6 6.6 6.6 1.4-1.4-6.6-6.6L24 9.4z"
                />
              </svg>
            </button>
          </div>
          <div
            class="bx--modal bx--bmrg-modal-confirm-container"
            role="presentation"
            tabindex="-1"
          >
            <div
              class="bx--modal-container"
            >
              <div
                class="bx--modal-header"
              >
                <p
                  class="bx--modal-header__heading bx--type-beta"
                >
                  Close this?
                </p>
                <button
                  class="bx--modal-close"
                  title="Close"
                  type="button"
                >
                  <svg
                    aria-label="Close"
                    class="bx--modal-close__icon"
                    focusable="false"
                    height="20"
                    preserveAspectRatio="xMidYMid meet"
                    role="img"
                    style="will-change: transform;"
                    viewBox="0 0 32 32"
                    width="20"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M24 9.4L22.6 8 16 14.6 9.4 8 8 9.4l6.6 6.6L8 22.6 9.4 24l6.6-6.6 6.6 6.6 1.4-1.4-6.6-6.6L24 9.4z"
                    />
                  </svg>
                </button>
              </div>
              <div
                id="confirm-modal-container"
              >
                <div
                  class="bx--modal-content"
                >
                  <p
                    class="bx--modal-content__text"
                  >
                    <div>
                      Your input will not be saved
                    </div>
                  </p>
                </div>
              </div>
              <div
                class="bx--modal-footer"
                inputref="[object Object]"
              >
                <button
                  class="bx--btn bx--btn--secondary"
                  id="bx--bmrg-confirm-modal-secondary-button"
                  tabindex="0"
                  type="button"
                >
                  NO
                </button>
                <button
                  class="bx--btn bx--btn--primary"
                  id="bx--bmrg-confirm-modal-primary-button"
                  tabindex="0"
                  type="button"
                >
                  YES
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
